### ISSUE TITLE
Slow Training Speed & Low GPU Utilization

### ISSUE THREAD

**Issue by donglabimaging** [DeepLabCut/DeepLabCut#362]
**Your operating system and DeepLabCut version**

Please state your operating system, env, and which version of DeepLabCut you are using:

Windows 10, with an Anaconda Env, & DeepLabCut 2.0.7

Please complete the following information about your system:

OS: Windows 10
DeepLabCut Version: 2.0.7
Browser: chrome
GPU: NVIDIA GeForce RTX 2070

**Describe the problem**

We have recently started using GPU for DeepLabCut, however, we noticed that the training takes extremely long. A 20 sec video of frame size 640x480 took more than 12 hours before DeepLabCut told us that the training was done and it was ready for evaluation. Although it seemed like the speed somewhat increased towards the end, in the beginning each iteration was taking about 5 minutes. We checked our computer's task manager during training, and it showed that the CPU was 15-20% engaged and the GPU only about 0-1%. But the dedicated GPU memory was about 7.2/8.0 GB, and this contrast seemed weird to us.

Originally, we thought the GPU was not being properly engaged. But we tried every tip we could find online:

1)
NVIDIA-SMI
(Displays details about our GPU)

2)
`tf.test.is_gpu_available()`
(Returns true)

3)
`tf.test.is_built_with_cuda()`
(Returns true)

4)
`tf.test.gpu_device_name()`
(Returns '/device:GPU:0')

5)
`with tf.device('/device:GPU:0'):`
     `deeplabcut.train_network(path_config_file, gputouse=0)`

6)
`from tensorflow.python.client import device_lib`
     `print(device_lib.list_local_devices())`
(Prints information about both CPU and GPU)

7)
`a = tf.constant([1.0, 2.0, 3.0, 4.0, 5.0, 6.0], shape=[2, 3], name='a')`
`b = tf.constant([1.0, 2.0, 3.0, 4.0, 5.0, 6.0], shape=[3, 2], name='b')`
`c = tf.matmul(a, b)`

`config = tf.ConfigProto(log_device_placement=True)`
`config.gpu_options.allow_growth = True`
`sess = tf.Session(config=config)`
`print(sess.run(c))`
(Printed output and Anaconda show that a, b, and c were all done using GPU)

8)
`import os`
`os.environ["CUDA_DEVICE_ORDER"]="PCI_BUS_ID"`
`os.environ["CUDA_VISIBLE_DEVICES"]="0"  # specify which GPU(s) to be used`
(Run at the very beginning)

These all seemed to indicate to us that the GPU was ready and set to go, but DeepLabCut's training was still extremely slow and the task manager still displayed the same kind of performance details. Interestingly, though, we tried the training with both the dlc-windowsCPU and dlc-windowsGPU environments, and the performance was significantly slower in the CPU environment (with CPU being engaged up to about 90%), so we suspect the GPU must be making some kind of a difference. Still, the slow speed in the GPU environment is concerning to us, especially when we can't find any other complaints like this online.

We read online that a low GPU utilization could be due to a small batch size, and we saw this comment by @MMathisLab on another post [https://github.com/AlexEMG/DeepLabCut/issues/224#issuecomment-468439622](url) saying that DeepLabCut's training batch size is 1 and that "in terms of GPU size, frame size is your limitation". This (we think) is also echoed in the paper "On the inference speed and video-compression robustness of DeepLabCut", so we are wondering if the only solutions to increasing speed are changing the batchsize and/or downsampling our videos.

**How to Reproduce the problem**
1. Activate dlc-windowsGPU on Anaconda
2. Open "Demo_yourowndata.ipynb" in Jupyter Notebook
3. In the training phase, run the following:
`with tf.device('/device:GPU:0'):`
`deeplabcut.train_network(path_config_file, gputouse=0)`
4. Check the performance

**Screenshots**

**Additional context**
We ran a simple autoencoder that uses keras and its training was super fast with GPU.

**Comment by MMathisLab** [comment]
I do not understand your bug report: 

> A 20 sec video of frame size 640x480 took more than 12 hours before DeepLabCut told us that the training was done

DLC doesn't use videos to train on, it uses frames. You input individual frames that you labeled, not videos. So, "20 sec video" doesn't mean anything to me. Also, as we say in multiple places, the size of your network scales with the size of your frames. Thus, if you have large frames, training will take longer. 

There is a timing section in the protocol paper: https://www.nature.com/articles/s41596-019-0176-0#Sec36

> The time required to train the network (Stage VII) mainly depends on the frame size of the dataset and the computer hardware. On an NVIDIA GeForce GTX 1080 Ti GPU, it takes 4 h to train the network for 200,000 iterations (640 × 480 pixels). On a CPU, it will take several days to train for the same number of iterations on the same dataset.

> Step 9, Stage VII, training the network: ~1–12 h (~6 h on a 1080-Ti GPU, and dependent on frame size)

You don't report how many iterations you ran, but 12hrs seems the appropriate amount of time given your frame size of 640x480. 

Seems you are indeed using up all the memory in your GPU card (the task manager is likely just reporting the GPU for your monitor, nvidia-smi is showing you the use of the card for python3): 

> But the dedicated GPU memory was about 7.2/8.0 GB, and this contrast seemed weird to us.

And, yes, of course a simple autoencoder is not a comparable speed to a ResNet-50.... 


>Additional context
We ran a simple autoencoder that uses keras and its training was super fast with GPU.

The idea is to not be making new models all the time, but building a robust network that you can use for video analysis, so a 1 or 2 time 12-hr run seems acceptable. Increasing batchsize > 1 is not supported in training at this time (yet this may change in a future release). 

You may consider looking at other benchmarks:

https://www.tensorflow.org/guide/performance/benchmarks
https://lambdalabs.com/blog/2080-ti-deep-learning-benchmarks/

But, a 2070 is generally not very fast.

**Comment by AlexEMG** [comment]
1% GPU use is weird. Can you post your logfile.txt? And output of nvidia-smi during training?

_A 20 sec video of frame size 640x480 took more than 12 hours before DeepLabCut told us that the training was done and it was ready for evaluation._

Whenever the loss flattens training can be stopped... This can be after 50-200k iterations for simple datasets.

**Comment by donglabimaging** [comment]
Dear @MMathisLab and @AlexEMG,

Thank you for the thoughtful responses. We are complete amateurs with these kinds of things, so we apologize for not being knowledgeable enough.

If we understood correctly, it seems then that the first few times of training the network are supposed to take very long, but as we keep refining it to improve analysis performance, the training time should gradually decrease; so the situation we described is to be expected.

Is the log.txt file attached below what you were referring to? If so, we took this from our first time training the network (iteration-0, under the "train" folder):
[log.txt](https://github.com/AlexEMG/DeepLabCut/files/3403727/log.txt)

We started retraining the network (ongoing as of now) after augmenting the training dataset with extracted outlier frames from a new but similar type of video, and here is the NVIDIA-SMI output:

![NVIDIA_SMI_Screenshot](https://user-images.githubusercontent.com/50585879/61409231-ca66f900-a8af-11e9-89ea-c0009036885c.png)

And we now understand that we can terminate the training after 50-200K iterations as the loss plateaus, thank you for the suggestion. That should cut back the training time significantly. We might also try downsampling our videos to make the frame size smaller.

Again, thank you very much!

**Comment by AlexEMG** [comment]
Looks good, the GPU is engaged! Indeed retraining with some new/additional frames should take <1h. Please make sure you initialize from the previous snapshot.


________________________________

**Comment by donglabimaging** [comment]
By that, do you mean we should change the init_weights parameter to, for us for example, "snapshot-1030000" (along with the full file path) in the "pose_cfg.yaml" file found in the "train" folder of the most recent iteration under "dlc-models"? Could you perhaps tell us if the following constitutes a proper flow of steps when retraining the network:

(After analyzing a novel video on the old network and examining the labeled video and seeing that there still needs improvement)

1. Extract outlier frames from that new video
2. Refine labels & check them
3. Merge the datasets
4. Create a new iteration of the training dataset
5. Add the file path of the new video (since it's now used for training) in "config.yaml"
6. Edit "init_weights" in the new "pose_cfg.yaml"
6. Train the network

Thank you!

**Comment by AlexEMG** [comment]
Yes, that’s the right workflow. (Note that merge creates the new iteration)

________________________________

**Comment by AlexEMG** [comment]
1. Extract outlier frames from that new video
2. Refine labels & check them
3 Merge the datasets
4. Create a new training split
5. Edit "init_weights" in the new "pose_cfg.yaml" pointing to previous snapshot
6. Train the network


‘Add the file path of the new video (since it's now used for training) in "config.yaml"’ this will happen during 1 automatically...

________________________________

**Comment by donglabimaging** [comment]
Thank you so much for helping us, we really appreciate it!

(And we don't know why but we were getting the "AUTOMATIC ADDING OF VIDEO TO CONFIG FILE FAILED! You need to do this manually for including it in the config.yaml file!" output in the extraction phase so we decided to add it manually.)
